packages/imperative/src/events/src/EventProcessor.ts

```diff
@@ -0,0 +1,174 @@
+ * - **Shared Events**: Zowe events that when triggered, notify all subscribed users.
+ * - **User Events**: Zowe events that are specific to only one user.
+ * - **Custom Events**: Applications can define their own shared and user events.
+ * @export
+ * @class EventProcessor
+ */
+export class EventProcessor {
+    public subscribedEvents: Map<string, Event> = new Map();
+    public eventTimes: Map<string, string>;
+    public processorType: IProcessorTypes;
+    public appName: string;
+    public logger: Logger;
+
+    /**
+     * Constructor initializes a new instance of EventProcessor.
+     *
+     * @param {string} appName - The application's name.
+     * @param {IProcessorTypes} type - The type of processor (Emitter, Watcher, or Both).
+     * @param {Logger} [logger] - Optional logger for recording events and errors.
+     */
+    public constructor(appName: string, type: IProcessorTypes, logger?: Logger) {
+        EventUtils.validateAppName(appName);
+
+        this.subscribedEvents = new Map();
+        this.eventTimes = new Map();
+        this.appName = appName;
+        this.processorType = type;
+
+        // Ensure correct environmental conditions to setup a custom logger,
+        // otherwise use default logger
+        if (ImperativeConfig.instance.loadedConfig == null || LoggerManager.instance.isLoggerInit === false) {
+            ConfigUtils.initImpUtils("zowe");
+        }
+        this.logger = logger ?? Logger.getImperativeLogger();
+    }
+
+    /**
+     * Subscription to an event that will notify all subscribed users.
+     *
+     * @param {string} eventName - The name of the event to subscribe to.
+     * @param {EventCallback[] | EventCallback} callbacks - Callback functions to handle the event.
+     * @returns {IEventDisposable} - Object allowing management/cleanup of the subscription.
+    */
+    public subscribeShared(eventName: string, callbacks: EventCallback[] | EventCallback): IEventDisposable {
+        if (this.processorType === IProcessorTypes.EMITTER) {
+            throw new ImperativeError({ msg: `Processor does not have correct permissions: ${eventName}` });
+        }
+        const isZoweEvent = EventUtils.isSharedEvent(eventName);
+        const eventType = isZoweEvent ? EventTypes.ZoweSharedEvents : EventTypes.SharedEvents;
+        const disposable = EventUtils.createSubscription(this, eventName, eventType);
+        EventUtils.setupWatcher(this, eventName, callbacks);
+        return disposable;
+    }
+
+    /**
+     * Subscription to an event that will notify a single user.
+     *
+    * @param {string} eventName - The name of the event to subscribe to.
+    * @param {EventCallback[] | EventCallback} callbacks - Callback functions to handle the event.
+    * @returns {IEventDisposable} - Object allowing management/cleanup of the subscription.
+    */
+    public subscribeUser(eventName: string, callbacks: EventCallback[] | EventCallback): IEventDisposable {
```

Comment:
It looks like lines 77-84 share a similar pattern to the lines for `subscribeUser`. 

Could duplication be minimized by creating a common, private `subscribe` function that takes in the type of event - then handling the event type accordingly?